File: dbf-mobile-host/ios/DBFMobileHost/DBFBridge.swift
import Foundation
import WebKit

final class DBFBridge: NSObject, WKScriptMessageHandler {
  private weak var webView: WKWebView?
  private let storage = StorageService()
  private let secure = SecureStorageService()
  private let device = DeviceService()
  private let notifications = NotificationService()

  private let allowedMethods: Set<String> = [
    "bridge.ready",
    "native.log",
    "storage.get",
    "storage.set",
    "storage.remove",
    "secure.get",
    "secure.set",
    "secure.remove",
    "device.info",
    "device.locale",
    "device.network",
    "deeplink.getInitial",
    "notifications.requestPermission",
    "notifications.getToken"
  ]

  init(webView: WKWebView) {
    self.webView = webView
  }

  func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
    guard let body = message.body as? String else { return }
    handleMessage(body)
  }

  private func handleMessage(_ raw: String) {
    guard let data = raw.data(using: .utf8) else { return }
    guard
      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      let type = json["type"] as? String,
      type == "invoke",
      let id = json["id"] as? String,
      let method = json["method"] as? String
    else {
      return
    }

    if !allowedMethods.contains(method) {
      sendResult(id: id, ok: false, payload: nil, error: ["code": "E_METHOD", "message": "Not allowed"])
      return
    }

    let payload = json["payload"] as? [String: Any]

    switch method {
    case "bridge.ready":
      sendEvent(name: "ready.ack", payload: ["ok": true])
      sendResult(id: id, ok: true, payload: ["ready": true], error: nil)
    case "native.log":
      sendResult(id: id, ok: true, payload: ["logged": true], error: nil)
    case "storage.get":
      let key = payload?["key"] as? String ?? ""
      let value = storage.get(key)
      sendResult(id: id, ok: true, payload: ["value": value as Any], error: nil)
    case "storage.set":
      let key = payload?["key"] as? String ?? ""
      let value = payload?["value"] as? String ?? ""
      storage.set(key, value: value)
      sendResult(id: id, ok: true, payload: ["success": true], error: nil)
    case "storage.remove":
      let key = payload?["key"] as? String ?? ""
      storage.remove(key)
      sendResult(id: id, ok: true, payload: ["success": true], error: nil)
    case "secure.get":
      let key = payload?["key"] as? String ?? ""
      let value = secure.get(key)
      sendResult(id: id, ok: true, payload: ["value": value as Any], error: nil)
    case "secure.set":
      let key = payload?["key"] as? String ?? ""
      let value = payload?["value"] as? String ?? ""
      secure.set(key, value: value)
      sendResult(id: id, ok: true, payload: ["success": true], error: nil)
    case "secure.remove":
      let key = payload?["key"] as? String ?? ""
      secure.remove(key)
      sendResult(id: id, ok: true, payload: ["success": true], error: nil)
    case "device.info":
      sendResult(id: id, ok: true, payload: device.info(), error: nil)
    case "device.locale":
      sendResult(id: id, ok: true, payload: device.locale(), error: nil)
    case "device.network":
      sendResult(id: id, ok: true, payload: device.networkStatus(), error: nil)
    case "deeplink.getInitial":
      let url = BridgeHub.shared.initialDeeplink
      BridgeHub.shared.initialDeeplink = nil
      sendResult(id: id, ok: true, payload: ["url": url as Any], error: nil)
    case "notifications.requestPermission":
      notifications.requestPermission { granted in
        if granted {
          self.notifications.registerForRemoteNotifications()
        }
        self.sendResult(id: id, ok: true, payload: ["granted": granted], error: nil)
      }
    case "notifications.getToken":
      let token = BridgeHub.shared.pushToken ?? ""
      sendResult(id: id, ok: true, payload: ["token": token, "platform": "ios"], error: nil)
    default:
      sendResult(id: id, ok: false, payload: nil, error: ["code": "E_UNKNOWN", "message": "Unknown method"])
    }
  }

  func sendEvent(name: String, payload: [String: Any]) {
    let event: [String: Any] = [
      "type": "event",
      "name": name,
      "payload": payload
    ]
    postToWebView(event)
  }

  private func sendResult(id: String, ok: Bool, payload: [String: Any]?, error: [String: Any]?) {
    var result: [String: Any] = [
      "type": "result",
      "id": id,
      "ok": ok
    ]
    if let payload = payload {
      result["payload"] = payload
    }
    if let error = error {
      result["error"] = error
    }
    postToWebView(result)
  }

  private func postToWebView(_ message: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: message),
          let json = String(data: data, encoding: .utf8) else { return }
    let script = "window.dispatchEvent(new MessageEvent('message', { data: \(json) }));"
    webView?.evaluateJavaScript(script, completionHandler: nil)
  }
}

File: dbf-mobile-host/android/app/src/main/java/com/dbf/mobile/DBFBridge.kt
package com.dbf.mobile

import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.google.firebase.messaging.FirebaseMessaging
import org.json.JSONObject

class DBFBridge(private val webView: WebView) {
  private val allowedMethods = setOf(
    "bridge.ready",
    "native.log",
    "storage.get",
    "storage.set",
    "storage.remove",
    "secure.get",
    "secure.set",
    "secure.remove",
    "device.info",
    "device.locale",
    "device.network",
    "deeplink.getInitial",
    "notifications.requestPermission",
    "notifications.getToken"
  )

  @JavascriptInterface
  fun postMessage(raw: String) {
    try {
      val json = JSONObject(raw)
      if (json.optString("type") != "invoke") return
      val id = json.optString("id")
      val method = json.optString("method")
      if (!allowedMethods.contains(method)) {
        sendResult(id, false, null, error("E_METHOD", "Not allowed"))
        return
      }
      when (method) {
        "bridge.ready" -> {
          sendEvent("ready.ack", JSONObject().put("ok", true))
          sendResult(id, true, JSONObject().put("ready", true), null)
        }
        "native.log" -> {
          Log.d("DBFBridge", json.optJSONObject("payload")?.toString() ?: "")
          sendResult(id, true, JSONObject().put("logged", true), null)
        }
        "storage.get" -> {
          val key = json.optJSONObject("payload")?.optString("key") ?: ""
          val prefs = getPrefs()
          val value = prefs.getString(key, null)
          val payload = JSONObject().put("value", value ?: JSONObject.NULL)
          sendResult(id, true, payload, null)
        }
        "storage.set" -> {
          val key = json.optJSONObject("payload")?.optString("key") ?: ""
          val value = json.optJSONObject("payload")?.optString("value") ?: ""
          getPrefs().edit().putString(key, value).apply()
          sendResult(id, true, JSONObject().put("success", true), null)
        }
        "storage.remove" -> {
          val key = json.optJSONObject("payload")?.optString("key") ?: ""
          getPrefs().edit().remove(key).apply()
          sendResult(id, true, JSONObject().put("success", true), null)
        }
        "secure.get" -> {
          val key = json.optJSONObject("payload")?.optString("key") ?: ""
          val value = getSecurePrefs().getString(key, null)
          val payload = JSONObject().put("value", value ?: JSONObject.NULL)
          sendResult(id, true, payload, null)
        }
        "secure.set" -> {
          val key = json.optJSONObject("payload")?.optString("key") ?: ""
          val value = json.optJSONObject("payload")?.optString("value") ?: ""
          getSecurePrefs().edit().putString(key, value).apply()
          sendResult(id, true, JSONObject().put("success", true), null)
        }
        "secure.remove" -> {
          val key = json.optJSONObject("payload")?.optString("key") ?: ""
          getSecurePrefs().edit().remove(key).apply()
          sendResult(id, true, JSONObject().put("success", true), null)
        }
        "device.info" -> {
          val payload = JSONObject()
            .put("platform", "android")
            .put("version", android.os.Build.VERSION.RELEASE ?: "")
            .put("model", android.os.Build.MODEL ?: "")
          sendResult(id, true, payload, null)
        }
        "device.locale" -> {
          val payload = JSONObject().put("locale", java.util.Locale.getDefault().toString())
          sendResult(id, true, payload, null)
        }
        "device.network" -> {
          val online = NetworkUtil.isOnline(BridgeRegistry.appContext)
          val payload = JSONObject().put("online", online)
          sendResult(id, true, payload, null)
        }
        "deeplink.getInitial" -> {
          val url = BridgeRegistry.initialDeeplink
          BridgeRegistry.initialDeeplink = null
          val payload = JSONObject().put("url", url ?: JSONObject.NULL)
          sendResult(id, true, payload, null)
        }
        "notifications.requestPermission" -> {
          val activity = BridgeRegistry.activity
          if (activity == null) {
            sendResult(id, false, null, error("E_CONTEXT", "No activity"))
            return
          }
          PermissionUtil.requestNotification(activity) { granted ->
            sendResult(id, true, JSONObject().put("granted", granted), null)
          }
        }
        "notifications.getToken" -> {
          FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            if (!task.isSuccessful) {
              sendResult(id, false, null, error("E_FCM", task.exception?.message ?: "Token error"))
              return@addOnCompleteListener
            }
            val token = task.result ?: ""
            BridgeRegistry.pushToken = token
            val payload = JSONObject().put("token", token).put("platform", "android")
            sendResult(id, true, payload, null)
          }
        }
        else -> sendResult(id, false, null, error("E_UNKNOWN", "Unknown method"))
      }
    } catch (_: Exception) {
      // ignore malformed payloads in scaffold
    }
  }

  private fun sendResult(id: String, ok: Boolean, payload: JSONObject?, error: JSONObject?) {
    val result = JSONObject()
      .put("type", "result")
      .put("id", id)
      .put("ok", ok)
    if (payload != null) result.put("payload", payload)
    if (error != null) result.put("error", error)
    postToWebView(result)
  }

  fun sendEvent(name: String, payload: JSONObject) {
    val event = JSONObject()
      .put("type", "event")
      .put("name", name)
      .put("payload", payload)
    postToWebView(event)
  }

  private fun postToWebView(message: JSONObject) {
    val json = message.toString()
    val script = "window.dispatchEvent(new MessageEvent('message', { data: $json }));"
    webView.post { webView.evaluateJavascript(script, null) }
  }

  private fun getPrefs() = webView.context.getSharedPreferences("dbf_prefs", 0)

  private fun getSecurePrefs(): android.content.SharedPreferences {
    val context = BridgeRegistry.appContext ?: webView.context
    val masterKey = MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build()
    return EncryptedSharedPreferences.create(
      context,
      "dbf_secure",
      masterKey,
      EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
      EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )
  }

  private fun error(code: String, message: String): JSONObject {
    return JSONObject().put("code", code).put("message", message)
  }
}
